oauth2client/flask.py

```diff
@@ -0,0 +1,329 @@
+
+        if not self.client_id or not self.client_secret:
+            if client_secrets_file:
+                self.client_id, self.client_secret = self._load_client_secrets(
+                    client_secrets_file)
+            elif 'OAUTH2_CLIENT_SECRETS_JSON' in app.config:
+                self.client_id, self.client_secret = self._load_client_secrets(
+                    app.config['OAUTH2_CLIENT_SECRETS_JSON'])
+            else:
+                self.client_id, self.client_secret = self._load_app_config(app)
+
+        app.register_blueprint(self._create_blueprint())
+
+        self.authorize_callback = authorize_callback
+        self.logout_callback = logout_callback
+
+        app.after_request(self._update_session)
+
+    def _load_client_secrets(self, filename):
+        client_type, client_info = clientsecrets.loadfile(filename)
+        if client_type != clientsecrets.TYPE_WEB:
+            raise ValueError(
+                "The flow specified in %s isn't supported." % client_type)
+
+        return client_info['client_id'], client_info['client_secret']
+
+    def _load_app_config(self, app):
+        if ('OAUTH2_CLIENT_SECRET' not in app.config or
+                'OAUTH2_CLIENT_ID' not in app.config):
+            raise ValueError(
+                "OAuth2 configuration could not be found. Either "
+                "OAUTH2_CLIENT_SECRETS_JSON or OAUTH2_CLIENT_ID and "
+                "OAUTH2_CLIENT_SECRET must be specified.")
+
+        return (app.config['OAUTH2_CLIENT_ID'],
+                app.config['OAUTH2_CLIENT_SECRET'])
+
+    def _update_session(self, resp):
+        """Ensures that the session's copy of the credentials is updated
+        when the credential refreshes its access token."""
+        ctx = _app_ctx_stack.top
+        if hasattr(ctx, 'oauth2_credentials'):
+            session['oauth2credentials'] = ctx.oauth2_credentials.to_json()
+        return resp
+
+    def _request_user_info(self):
+        """Gets the user's profile information from Google+
+
+        Makes an HTTP request to the Google+ API to retrieve the user's basic
+        profile information, including full name and photo.
+        """
+        http = self.http()
+        resp, content = http.request(
+            'https://www.googleapis.com/plus/v1/people/me')
+
+        if resp.status != 200:
+            current_app.logger.error(
+                "Error while obtaining user profile: %s" % resp)
+            return None
+
+        try:
+            profile = json.loads(content)
+        except ValueError:
+            current_app.logger.error("Invalid user profile json: %s", content)
+            return None
+
+        return profile
+
+    def _make_flow(self):
+        # Generate a CSRF token to prevent malicious requests.
+        state = ''.join(random.choice(string.ascii_uppercase + string.digits)
+                        for _ in range(32))
+
+        session['oauth2state'] = state
```

Comment:
The pattern I'm recommending elsewhere is to store both this token and the 'continuation url' in the state. In Java I put an encoded blob of JSon in here in case we need more than these 2 fields. If the continuatino URL is not in here, how are you getting back to the original page? Session State?